### PR TITLE
Glue example notebook

### PR DESCRIPTION
**Pull request by ishanashastri**
This PR adds a notebook that serves as documentation for the new GLUE entrypoint. The notebook covers the purpose of the entrypoint, how to construct your `hparams`, and demonstrates running a fine-tuning job from a pre-trained checkpoint.

### CODE REVIEW

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:9]:
```suggestion
    "This notebook is intended to demonstrate how to use the [GLUE (General Language Understanding Evaluation)](https://gluebenchmark.com/) entry point for pre-training and fine-tuning NLP models across the 8 GLUE tasks.\n This will cover:\n",
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:30]:
```suggestion
    "This tutorial requires access to an AWS S3 Bucket. If AWS credentials are not already available in your environment, then you will need to obtain an `AWS_ACCESS_KEY_ID` and an `AWS_SECRET_ACCESS_KEY` that have permission to upload to and download from a S3 Bucket.\nTo learn more about AWS S3 and how to configure it check out [AWS S3 documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/GetStartedWithS3.html)."
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:61]: Just checking: this doesn't require mosaicml[nlp]?

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:70]:
```suggestion
    "This entrypoint allows you to specify if you want to pre-train a NLP model, fine-tune a model on the downstream tasks, or run the entire pipeline. If pre-training, the entry point will handle distributed training across all available GPUs. If fine-tuning, the entry point will fine-tune all given checkpoints on all 8 GLUE tasks in parallel using multiprocessing pools. This entry point is designed to make this process more efficient and remove the tediousness of individually spawning jobs and manually loading all the model checkpoints. "
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:79]:
```suggestion
    "A full out-of-the-box YAML example for this entry point can be found in `./glue_example.yaml.` If you're already familiar with YAMLs, you can skip to the next part! If not, we'll break down how this is structured."
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:187]:
```suggestion
    "If you are only fine-tuning checkpoints on the GLUE tasks, you are expected to specify the checkpoints to load from by specifying a `finetune_ckpts` list as so in the `finetune_hparams` section of your YAML. Upon running the entry point with this list, it will automatically pull all checkpoints and fine-tune on all of them. Note that if the `finetune_ckpts` list contains paths in object store, the entry point expects an `ObjectStoreLogger` and `load_object_store` instances, as well as their corresponding credentials to be specified, otherwise it will try to load from local disk. See our [checkpointing guide](https://docs.mosaicml.com/en/stable/trainer/checkpointing.html#resume-training) if you're not familiar with our checkpoint saving and loading schema. \n",
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:228]:
```suggestion
    "To run the entire end-to-end pipeline, you are expected to provide the entry point with your pre-train configuration as explained above, as well as any overrides to apply to the fine-tuning jobs. In this case, the entry point is run in two distinct stages for distributed pre-training and multiprocessed fine-tuning; however, all information transferred between the stages is automatically handled by the entry point.  Checkpoints are automatically saved to your specified `save_folder` and loaded from wherever pre-training saved them, therefore the `finetune_ckpts` section of `finetune_hparams` is ignored if specified."
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:235]:
```suggestion
    "❗ Note: The entry point runs all 8 GLUE fine-tuning tasks on every saved pre-training checkpoint, so set your [save_interval](https://docs.mosaicml.com/en/stable/trainer/checkpointing.html?highlight=sve%20interval#save-interval) within your `pretrain_hparams` appropriately to avoid unnecessarily long evaluation times. "
```

Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:235]:
```suggestion
    "❗ Note: The entrypoint runs all 8 GLUE fine-tuning tasks on every saved pre-training checkpoint, so set your [save_interval](https://docs.mosaicml.com/en/stable/trainer/checkpointing.html#save-interval) within your `pretrain_hparams` appropriately to avoid unnecessarily long evaluation times. "
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:244]:
```suggestion
    "Let's now put together all our knowledge about the entry point and launch a job that will fine-tune a pre-trained BERT model on the 8 GLUE tasks! Because we are only fine-tuning with no special configurations, we only need to specify our bucket information and the `finetune_ckpts` to load from:"
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:268]:
```suggestion
    "Let's now dump our constructed hparams to a YAML file to be loaded by the entry point:"
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:319]: When running this in Colab I get an error:
`can't open file 'examples/glue/run_glue_trainer.py': [Errno 2] No such file or directory`
I suspect this is because the notebook only runs pip install, but does not pull the core itself?

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:331]:
```suggestion
    "❗ Note: You must run this entry point via the **python** command, not the composer one!\n",
```

Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:331]: I am not sure this note is important, since you have the command using python above.

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:333]:
```suggestion
    "💡 Pro-tip: Try `python examples/glue/run_glue_trainer.py --help` to get more information about the entry point, and `python examples/glue/run_glue_trainer.py {pretrain_hparams, finetune_hparams} --help` to get a detailed breakdown of your hparams options!"
```

File: examples/glue/glue_entrypoint.ipynb
Comment (lupesko) [inline examples/glue/glue_entrypoint.ipynb:342]:
```suggestion
    "Try pre-training and fine-tuning your own models with this framework! Also, feel free to check out the rest of our [Composer docs](https://docs.mosaicml.com/) to try using Composer speedups in this entry point!\n"
```

**Review by lupesko — CHANGES_REQUESTED** [review]
Left a bunch of comments, most of them are simple nits for wording and syntax.

Critical comments:
(1) The notebook failed to run the entry point for me when running on Colab... I suspect that the notebook assumes I have the code locally, however just running pip install does not pull the GH repo. This can be addressed by either: (1) Making the entry point a command packaged into the pip package (2) Adding a command to pull the repo
(2) Requiring AWS account may be a blocker for folks. Is this mandatory? can we have the example not require that?

File: examples/glue/glue_entrypoint.ipynb
Comment (ishanashastri) [reply examples/glue/glue_entrypoint.ipynb]: Good catch, added git cloning!

**Comment by lupesko** [comment]
It's now running for me on Colab, although it fails probably due to lacking AWS access.
@ishanashastri :
- is the plan to update the notebook without requiring aws access once #1390 is merged?
- Jenkins is failing, please see if the issue is related to the new notebook

**Comment by ishanashastri** [comment]
> * is the plan to update the notebook without requiring aws access once [Decouple GLUE entry point saving and loading logic  #1390](https://github.com/mosaicml/composer/pull/1390) is merged?

Yep, exactly. This will be blocked until that is merged, at which point I can refactor the AWS logic out. 

> * Jenkins is failing, please see if the issue is related to the new notebook

Will address after #1390 is merged!

**Review by lupesko — APPROVED** [review]

**Comment by lupesko** [comment]
Approved!

**Comment by ishanashastri** [comment]
Will not merge until #1390 is merged!

**Review by hanlint — APPROVED** [review]

**Pull request merged**
